distribution/addon_distribution.py

```diff
@@ -0,0 +1,201 @@
+            raise ValueError(f"{downloader_type} not implemented")
+        return downloader()
+
+    @classmethod
+    @abstractmethod
+    def download(cls, addon_url, destination):
+        """Returns url to downloaded addon zip file.
+
+        Args:
+            addon_url (str): http or OS or any supported protocol url to addon
+                zip file
+            destination (str): local folder to unzip
+        Retursn:
+            (str) local path to addon zip file
+        """
+        pass
+
+    @classmethod
+    def check_hash(cls, addon_path, addon_hash):
+        """Compares 'hash' of downloaded 'addon_url' file.
+
+        Args:
+            addon_path (str): local path to addon zip file
+            addon_hash (str): sha256 hash of zip file
+        Raises:
+            ValueError if hashes doesn't match
+        """
+        if not os.path.exists(addon_path):
+            raise ValueError(f"{addon_path} doesn't exist.")
+        if addon_hash != RemoteFileHandler.calculate_md5(addon_path):
+            raise ValueError(f"{addon_path} doesn't match expected hash.")
+
+    @classmethod
+    def unzip(cls, addon_zip_path, destination):
+        """Unzips local 'addon_zip_path' to 'destination'.
+
+        Args:
+            addon_zip_path (str): local path to addon zip file
+            destination (str): local folder to unzip
+        """
+        RemoteFileHandler.unzip(addon_zip_path, destination)
+        os.remove(addon_zip_path)
+
+    @classmethod
+    def remove(cls, addon_url):
+        pass
+
+
+class OSAddonDownloader(AddonDownloader):
+
+    @classmethod
+    def download(cls, addon_url, destination):
+        # OS doesnt need to download, unzip directly
+        if not os.path.exists(addon_url):
+            raise ValueError("{} is not accessible".format(addon_url))
+        return addon_url
+
+
+class HTTPAddonDownloader(AddonDownloader):
+    CHUNK_SIZE = 100000
+
+    @classmethod
+    def download(cls, addon_url, destination):
+        cls.log.debug(f"Downloading {addon_url} to {destination}")
+        file_name = os.path.basename(destination)
+        _, ext = os.path.splitext(file_name)
+        if (ext.replace(".", '') not
+                in set(RemoteFileHandler.IMPLEMENTED_ZIP_FORMATS)):
+            file_name += ".zip"
+        RemoteFileHandler.download_url(addon_url,
+                                       destination,
+                                       filename=file_name)
+
+        return os.path.join(destination, file_name)
+
+
+def get_addons_info(server_endpoint):
+    """Returns list of addon information from Server"""
+    # TODO temp
+    # addon_info = AddonInfo(
+    #     **{"name": "openpype_slack",
+    #        "version": "1.0.0",
+    #        "addon_url": "c:/projects/openpype_slack_1.0.0.zip",
+    #        "type": UrlType.FILESYSTEM,
+    #        "hash": "4f6b8568eb9dd6f510fd7c4dcb676788"})  # noqa
+    #
+    # http_addon = AddonInfo(
+    #     **{"name": "openpype_slack",
```

Comment:
Just thinking about this. I would say that single item is representing single addon version and the item should have something like `"sources"` where are the source items with `"type"` and type related data. I would expect that `type` is only required key and the type handler would define remaining required keys. At the end you don't want download each source but check if the addon is available locally and if not then try sources one by one with source type specific handlers.

Also the data are what server will return so it will contain only json compatible types.

Example:
```javascript
{
    "name": "openpype_slack",
    "version": "1.0.0",
    // Hash is probably same for all sources?
    "hash": "4f6b8568eb9dd6f510fd7c4dcb676788",
    "sources": [
        {
            "type": "http",
            "url": "https://some/path"
        },
        {
            "type": "filesystem",
            "path": {
                "windows": ["P:/sources/some_file.zip", "W:/sources/some_file.zip"],
                "linux": ["/mnt/srv/sources/some_file.zip"],
                "darwin": ["/Volumes/srv/sources/some_file.zip"]
            }
        },
        {
            "type": "filesystem",
            "path": "P:/sources/some_file.zip"
        }
    ]
}
```

Reply:
Updated format